Add precompile statements

lkdvos opened the pull request:
This PR is an attempt to fix some compilation time issues, as brought up by @amilsted.
The problem is the functions attached [in this file](https://gist.github.com/lkdvos/662adba553b9ed33ff95e59c95c0d25f)

## Current status
<details>
<summary>
(v5.1.3)
</summary>

**Code Loading + Precompilation**
```bash
❯ julia --project -e '@time using TensorOperations'
  0.016875 seconds (32.41 k allocations: 2.502 MiB)
  
❯ julia --project -e '@time Base.compilecache(Base.identify_package("TensorOperations"))'
  0.660362 seconds (18.10 k allocations: 1.061 MiB, 0.71% compilation time)
```
**TTFX**
```bash
❯ julia --project -e 'include("to_dense.jl"); @time to_dense_ncon(M, L, R)'
 17.228799 seconds (90.38 M allocations: 4.563 GiB, 2.30% gc time, 99.99% compilation time)
 
❯ julia --project -e 'include("to_dense.jl"); @time to_dense_tensor(M, L, R)'
 16.379553 seconds (80.78 M allocations: 4.093 GiB, 2.19% gc time, 100.00% compilation time)
```
**Performance**
```bash

❯ julia --project -e 'include("to_dense.jl"); using BenchmarkTools; @btime to_dense_ncon(M, L, R)'
  112.833 μs (825 allocations: 437.98 KiB)

❯ julia --project -e 'include("to_dense.jl"); using BenchmarkTools; @btime to_dense_tensor(M, L, R)'
  65.416 μs (56 allocations: 408.83 KiB)
```
</details>

## Precompilation

After investigating, there are several possible improvements.
Given that there does not seem to be any type instability issues, and I couldn't spot any obvious problems, I figured that this is partially just because we really do want to specialize quite a bit, assuming that TensorOperations is an important performance bottleneck.
Therefore, I started out by simply adding precompile statements:

<details>
<summary>
This PR: adding precompilation
</summary>

**Code Loading + Precompilation**
```bash
❯ julia --project -e '@time using TensorOperations'
  0.823528 seconds (689.87 k allocations: 68.223 MiB, 7.65% gc time)
  
❯ julia --project -e '@time Base.compilecache(Base.identify_package("TensorOperations"))'
137.465629 seconds (17.85 k allocations: 1.048 MiB, 0.00% compilation time)
```
**TTFX**
```bash
❯ julia --project -e 'include("to_dense.jl"); @time to_dense_ncon(M, L, R)'
  0.963772 seconds (8.88 M allocations: 444.266 MiB, 4.24% gc time, 99.87% compilation time)
 
❯ julia --project -e 'include("to_dense.jl"); @time to_dense_tensor(M, L, R)'
  0.059043 seconds (264.73 k allocations: 13.633 MiB, 99.53% compilation time)
```
**Performance**
```bash
❯ julia --project -e 'include("to_dense.jl"); using BenchmarkTools; @btime to_dense_ncon(M, L, R)'
  114.541 μs (825 allocations: 437.98 KiB)

❯ julia --project -e 'include("to_dense.jl"); using BenchmarkTools; @btime to_dense_tensor(M, L, R)'
  60.792 μs (56 allocations: 408.83 KiB)
```
</details>

The good: TTFX is negligible and the performance seems unaffected!
The bad: precompilation time now takes ~140 seconds...

## Standardization

Because this package often forms the important part of code for performance, this might be acceptable, but we can improve by standardizing some of the types in StridedViews, see [this PR](https://github.com/Jutho/StridedViews.jl/pull/15).
The precompile time is now at ~40 seconds, while TTFX is still good and performance is mostly unchanged. Interestingly though, that StridedView approach does require an additional allocation (`reshape` seems to cause this?), of which I'm unsure if we can get rid of it.

<details>
<summary>
This PR + StridedViews standardizing parenttype
</summary>

**Code Loading + Precompilation**
```bash
❯ julia --project -e '@time using TensorOperations'
  0.641844 seconds (223.34 k allocations: 19.020 MiB, 6.27% gc time)
  
❯ julia --project -e '@time Base.compilecache(Base.identify_package("TensorOperations"))'
 41.829745 seconds (17.85 k allocations: 1.048 MiB, 0.01% compilation time)
```
**TTFX**
```bash
❯ julia --project -e 'include("to_dense.jl"); @time to_dense_ncon(M, L, R)'
  1.001807 seconds (8.89 M allocations: 444.702 MiB, 4.45% gc time, 99.81% compilation time)
 
❯ julia --project -e 'include("to_dense.jl"); @time to_dense_tensor(M, L, R)'
  0.064136 seconds (270.55 k allocations: 13.911 MiB, 6.60% gc time, 99.63% compilation time)
```
**Performance**
```bash
❯ julia --project -e 'include("to_dense.jl"); using BenchmarkTools; @btime to_dense_ncon(M, L, R)'
  115.417 μs (844 allocations: 438.23 KiB)

❯ julia --project -e 'include("to_dense.jl"); using BenchmarkTools; @btime to_dense_tensor(M, L, R)'
  61.250 μs (75 allocations: 409.08 KiB)
```
</details>

As an interesting sidenote, without the precompilation statements here, but including the StridedViews standardization actually already cuts the TTFX quite considerably:

<details>
<summary>
only StridedViews standardizing parenttype
</summary>

**Code Loading + Precompilation**
```bash
❯ julia --project -e '@time using TensorOperations'
  0.017374 seconds (34.04 k allocations: 2.602 MiB)
  
❯ julia --project -e '@time Base.compilecache(Base.identify_package("TensorOperations"))'
  0.775875 seconds (18.09 k allocations: 1.060 MiB, 0.57% compilation time)
```
**TTFX**
```bash
❯ julia --project -e 'include("to_dense.jl"); @time to_dense_ncon(M, L, R)'
  9.672027 seconds (40.09 M allocations: 1.996 GiB, 1.98% gc time, 99.98% compilation time)
 
❯ julia --project -e 'include("to_dense.jl"); @time to_dense_tensor(M, L, R)'
  8.857590 seconds (31.01 M allocations: 1.552 GiB, 1.64% gc time, 100.00% compilation time)
```
**Performance**
```bash
❯ julia --project -e 'include("to_dense.jl"); using BenchmarkTools; @btime to_dense_ncon(M, L, R)'
  114.333 μs (844 allocations: 438.23 KiB)

❯ julia --project -e 'include("to_dense.jl"); using BenchmarkTools; @btime to_dense_tensor(M, L, R)'
  63.042 μs (75 allocations: 409.08 KiB)
```
</details>

Jutho commented:
I am certainly ok with this. There used to be precompile statements present, but then at some point removed them as the precompilationtools ecoystem was evolving rapidly and I couldn't follow up.

lkdvos commented:
I've updated the implementation and added a section to the documentation page to explain what is going on. In particular, we now follow the same convention as `PrecompileTools` for disabling precompilation, and the precompile settings can be tweaked locally for people that want this.
If everything turns green, I would say we can merge this and tag a new version. I would actually think that this might warrant a v5.2.0 release, what do you think?

amilsted commented:
Thanks @lkdvos, this looks like it helps quite a bit. There's still a lot of compile going on, but I understand you want to avoid runtime dispatch and allocating vectors for indices and such.

A general comment: There is quite a deep call stack of functions that pass `Index2Tuple` arguments from level to level. Because these args are tuples, most of this stack (tensorcontract, stridedtensorcontract, blascontract, ...) gets recompiled for every shape combination in a pairwise contraction! Is it worth allowing these arguments to be arrays, for example if they are passed from `ncon`, until we are deeper in the stack? It seems kind of nuts that in some of the MPS code I am looking at we have >200 specializations of `_blas_contract!` (taking ~8s to compile) according to SnoopCompile tools. Maybe that really does need specializing (to avoid index array allocs), but higher-level `tensorcontract!` methods also have 10-100 specializations (some of them almost 10s of cumulative compile time) and account for a negligible portion of the runtime. Many of these are thin wrappers, like the ones that specialize on conj args (but do nothing with the tuples). Seems like these could (optionally) pass vectors along that later get converted to tuples for specialization at a lower level.

Also, have you thought about allowing these tuples to be arrays more generally and using something like bumper to allocate vectors that come from index manipulating functions?

lkdvos commented:
@amilsted , does this include the changes from [the StridedViews PR](https://github.com/Jutho/StridedViews.jl/pull/15)? I was hoping that it is that part that mitigates the expensive compilations of `_blas_contract!`, while the `tensorcontract!` specializations actually don't matter too much.
I have to admit that I don't actually have much feeling with what causes the largest amount of compilation time. I somehow figured that the higher-up callers are thin wrappers, so compiling these functions should not be that expensive. On the other hand, the heavy-duty lower-level functions would be the expensive ones to compile (which is probably also what we want because these determine the performance?).

I don't think the allocations of the index arrays are really the problem. I think one of the main reasons we chose tuples instead of `Vector`s of indices, and convert to tuples as soon as possible is that this allows type-stable calls in the following:
```julia
tensorcontract(A, pA, conjA, B, pB, conjB, pAB, ...) -> C
```
Here a `Vector` would actually make the result `Array{T,N} where {N}` instead. (`N = numind(pAB)` is available at compile-time).
Our philosophy was to convert everything to tuples as high up the chain as possible to avoid having to recompile the entire chain, with the idea that the `Tuple` way is the most performant. (This is also why `@tensor` does not have this problem, since it immediately has `Tuple`s).
`ncon` is a bit annoying in that regard, since there none of this is known at compile-time, and therefore we pay the price of extra compilation without the benefit of type stability.
Obviously, this might have been a mistake, it's not that easy to say and I just wanted to give my reasoning.

One thing we could consider experimenting with as well is to simply put `@nospecialize` annotations around the `pA`, `pB` etc arguments, and investigate how that affects performance? It does tend to be a bit of a mess to start battling the compiler heuristics like this, but I do agree that the current compilation time is a bit excessive.

amilsted commented:
@lkdvos This is with the StridedViews PR. With this PR, it improves the MPO contraction case we looked at significantly, but I'm still seeing a lot of compile on code that does more MPS stuff (sweeping algorithms). I should say that I have measured *inference* time rather than LLVM compile time. I don't know why so much time is being spent on inference. Perhaps there's a lot of constant propagation happening that blows up the cases?

Jutho commented:
@amilsted , I won't have any time to look at this before at least Thursday, but if you could share some of these snoopcompile outputs or scripts, that would be a great help in analyzing if there are any obvious improvements up for grabs.

amilsted commented:
Ok, it looks like part of it was that the other code I looked at was using the Bumper allocator by default. Any chance of getting those methods precompiled too? :)

amilsted commented:
I will see if I can usefully share some snoopcompile outputs for what remains. There does indeed seem to be a lot of constant propagation happening (lots of orange-gold in plots like [this](https://timholy.github.io/SnoopCompile.jl/dev/tutorials/snoop_inference/#flamegraph)). Attempting to disable it at the ncon-caller level doesn't seem to get rid of it.

amilsted commented:
![flame](https://github.com/user-attachments/assets/56aeb6a2-9480-4faf-981a-dbb34d7b525b)

The golden-orange columns are TensorOps/Strided methods.

amilsted commented:
Yeah, fair enough. Perhaps we should just not default to using Bumper...

lkdvos commented:
I messed around with it a bit, and I did manage to do some (possibly slightly evil) trickery to at least slightly standardize some of the types on Julia 1.11+ by using explicit `Memory` objects.
I also included the changes from the StridedViews standardization immediately here, since I realized that `parent(::StridedView)` might actually be used, and people might have depended on `parent(StridedView(a)) === a`, which would no longer be the case.

I do really want to have the support for Bumper though, as it seems like a good feature.

I'll try around a bit more tomorrow to see if I can get a `PrecompileTools`-friendly way of setting all of this up, which would also allow downstream packages to more easily set up their own precompilation workloads.

amilsted commented:
Just tried the latest version of this branch. For the default allocator, TTFX for my DMRG test case is ~50s with Bumper and ~13s with the default allocator. Looking good in the latter case!

lkdvos reviewed (commented):
For me, this should now be good to go.

Summarizing:

- I added precompilation statements for regular arrays and bumper arrays (UnsafeArrays).
- precompilation settings can be tweaked locally through a preferences system, as explained in a new section of the documentation
- To reduce compilation times, I attempted to "normalize" types a bit: instead of using `StridedView(array)`, I now have a hook function `wrap_stridedview` that, whenever possible, simply returns a stridedview of the underlying `Memory` (julia +v1.11). This ensures that the actual performance critical sections do not have to be recompiled for many different combinations of types. (note that this cannot be done in StridedViews.jl, since this breaks `parent(StridedView(a)) === a`.)
- I removed an `@inline` before `makeblascontractable`, since that seems like a large function to force inlining. (not clear if this helps or hurts, but seemed reasonable to me)

@amilsted it would be nice if you could retry and tell me if the bumper precompilation helped at all.

lkdvos commented:
Never mind, it seems like the Bumper precompilation on 1.10 takes absolutely forever, so I might have to disable that...

amilsted commented:
Sorry, was busy at the March Meeting and stuff. Will check this PR against non-bumper version of the code, though I guess it will still show the improvements I already saw above. Even if you can't add bumper precomp in 1.10, I think this is still a great improvement.

Jutho commented:
Strided.jl 2.3 has been tagged; time to merge this?

lkdvos commented:
Seems like the bumper extension isn't precompiling, I'll try investigate Monday

lkdvos commented:
I can't seem to figure out why the `Bumper` extension takes that long to precompile. I've added some print statements to the precompilation part of the code, it seems like it gets through the loops just fine, but then hangs at the very end for about 10 mins on my mac, which seems to be enough to make the github runners time out/error.

In order to avoid stalling this PR forever, I simply deleted the precompilation for the Bumper extension for now. When I find some more time I can reinvestigate, but for now the precompilation of the rest should at least already be a great improvement.

The remaining issues should be solved by https://github.com/Jutho/StridedViews.jl/pull/16

lkdvos commented:
@Jutho, this is good to go

Jutho commented on src/implementation/blascontract.jl:
```diff
@@ -55,8 +55,8 @@ function _blas_contract!(C, A, pA, B, pB, pAB, α, β, backend, allocator)
-    flagA || tensorfree!(A_.parent, allocator)
-    flagB || tensorfree!(B_.parent, allocator)
+    flagA || tensorfree!(A_, allocator)
+    flagB || tensorfree!(B_, allocator)
```
What was the reason for this change? Will this still work if `A_` is a `StridedView` over a `PtrArray`? Or do we need a `tensorfree!(A::StridedView, allocator) = tensorfree!(A.parent, allocator)` definition?

lkdvos replied:
Oh I meant to revert this as well,  I changed things around a bit when I wanted to do the normalization of the Strided parent here, but I'll change it back tomorrow.

Jutho commented:
Look's great; thanks again. I'll merge.
